## Lambda Functions

A lambda operator or lambda function is used for creating small, one-time, anonymous function objects in Python.<br><br>
The general expression for a lambda function is <br>
lambda argument(s): expression <br><br>
A lambda function can take any number of arguments, but they contain only a single expression. An expression is a piece of code executed by the lambda function, which may or may not return any value. <br><br>

Let us look at some examples using lambda functions.

In [1]:
sum1 = lambda x, y, z: x + y + z

sum1(1,2,3)

6

In [2]:
sum1(2,3,4)

9

**Obtain the minimum among two numbers**

In [1]:
min_value = (lambda x, y: x if x < y else y)
min_value(34, 78)

34

**Check whether two strings are equal**

In [2]:
result = (lambda x, y: 'True' if x == y else 'False')
result("Data", "DaTa")

'False'

**Demonstrating wrong usage of variable name**

In [4]:
# multiplication - wrong usage of variable
x = lambda a, c : a + b
print(x(5, 6))

NameError: name 'b' is not defined

**If variable a and c are used, then the operations should be on a and c.**

### Overview of map, reduce and filter functions

The function map(f,l) applies a function 'f' to each element of list 'l' <br><br>
The output of map is not a list, to convert it to a list use list(map(f,l))<br><br>
The function filter(p,l) where property p is checked for every element in list l. The output of filter(p, l) is a sublist of list 'l' that satisfies the property 'p'. <br><br>
The reduce(fun,seq) function is used to apply a particular function passed in its argument to all of the list elements mentioned in the sequence passed along and returns only one value.This function is defined in the “functools” module.

### Function map()

**A test was conducted in a class, the list of all the scores is given. Tha scores are out of 45. Obtain the percentage of each score using lambda function.**
             
             marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]

In [11]:
marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]

x = lambda x: x * 100 / 45

print(list(map(x, marks)))

[44.44444444444444, 51.111111111111114, 75.55555555555556, 100.0, 73.33333333333333, 22.22222222222222, 95.55555555555556, 51.111111111111114, 60.0, 75.55555555555556, 82.22222222222223, 97.77777777777777, 93.33333333333333, 95.55555555555556]


In [ ]:
marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]

print(list(map(lambda x:x*100/45 , marks)))

In [4]:
marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]

print(list(map(lambda x, y=100 :x*y/45 , marks)))

[44.44444444444444, 51.111111111111114, 75.55555555555556, 100.0, 73.33333333333333, 22.22222222222222, 95.55555555555556, 51.111111111111114, 60.0, 75.55555555555556, 82.22222222222223, 97.77777777777777, 93.33333333333333, 95.55555555555556]


The output is the percentage scores.

**Create two lists and add corresponding elements of each list**

In [5]:
# iteration using list
list_a = [30, 35, 40]
list_b = [12, 5, 70]
  
print(list(map(lambda x, y: x + y, list_a, list_b)))

[42, 40, 110]


### Function filter()

**To obtain the  sum of cubes of odd number from 1 to 50**

In [6]:
# sum of cubes of odd numers from 1 to 50

# user defined function to obtain the cube
def cube(num):
    return(num*num)

In [7]:
# user defined function to obtain odds numbers from a list
def isodd(num):
    return(num%2 == 1)

In [15]:
even = lambda x: x % 2 == 0
even(3,4,5,67)

TypeError: <lambda>() takes 1 positional argument but 4 were given

In [16]:
list(map(lambda x: x % 2 == 0, range(1,20)))

[False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False]

In [12]:
list(filter(lambda x: x % 2 == 0, range(1,20)))

[2, 4, 6, 8, 10, 12, 14, 16, 18]

In [8]:
# create a list of all cubes
list_cubes = list(map(cube,filter(isodd,range(1,50))))

# sum of all elements in the in the list_cube
sum(list_cubes)

20825

In [9]:
# create a list of all cubes
list_cubes = list(map(cube,filter(isodd,range(1,50))))

# sum of all elements in the in the list_cube
sum(list_cubes)

20825

The same code above can be done using list comprehension  in the following way:

In [10]:
# map - the function defined above is called 
# generator - the range for numbers is fixed
# filter - cube is calculated for the numbers which satisfy the 'if' condition
# built-in function sum() is used to obtain the sum

sum ( [cube(x) for x in range(50) if isodd(x)] ) # Map - Generator - Filter

20825

**To obtain the odd numbers from a list**

In [11]:
numbers = [121, 443, 477, 125, 34, 655, 458, 141, 363, 125, 456 , 578, 123]

# function that filters the numbers
def oddnumber(in_num):
    if(in_num % 2) == 0:
        return True
    else:
        return False

In [12]:
# Use filter() to remove odd numbers
out_filter_number = filter(oddnumber, numbers)

print("Type of filter object: ", type(out_filter_number))
print("Even number is as follows: ", list(out_filter_number))

Type of filter object:  <class 'filter'>
Even number is as follows:  [34, 458, 456, 578]


### Function reduce()

**List of  sales in a shop are noted. Obtain the total sales**

In [14]:
# import reduce function from functools
from functools import reduce

sale = [14565,65460,74450,54546,23245,14543,84355,54566]
total_sale = reduce(lambda a, b: a + b, sale)
print(total_sale)

385730


**Thus we may say the total sale is 385730 INR.**

**Factorial of a 5** 

In [17]:
# import reduce function from functools
from functools import reduce

factorial = reduce(lambda p,q : p*q , range(1,6))

print("The factorial of 5 is ",factorial)

The factorial of 5 is  120


In [18]:
print(factorial)

120
